pandas/io/sql.py

```diff
@@ -97,6 +97,38 @@ def _handle_date_column(col, utc=None, format=None):
             return to_datetime(col, errors="coerce", format=format, utc=utc)
 
 
+def _is_datetime_series_with_different_offsets(df_col):
```

Comment:
I think `infer_dtype` can detect whether the dtype should be upcast to `datetimetz`. Otherwise the logic should live there.

Reply:
Or use `maybe_convert_objects`

Reply:
I have tested infer_dtype, maybe_convert_objects and is_datetime64tz_dtype, and my function _is_datetime_series_with_different_offsets. Only my function works. So we should keep it unless someone has a better idea.

<details>
<summary>Test of infert_dtype, maybe_convert_objects, is_datetime64tz_dtype, _is_datetime_series_with_different_offsets</summary>

```python
from pandas.api.types import infer_dtype
from pandas._libs.lib import maybe_convert_objects
from pandas.api.types import is_datetime64tz_dtype
from pandas.io.sql import _is_datetime_series_with_different_offsets
import pandas as pd
import psycopg2
import datetime
import numpy as np


# this is the kind of data we receive from a psycopg2 cursor when we have a timestamptz column in a postgres table
data_with_offsets = [datetime.datetime(2019, 11, 14, 16, 12, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60)),
                     datetime.datetime(2019, 8, 7, 15, 37, 4, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=120))]

data_without_offsets = [datetime.datetime(2019, 11, 14, 16, 12, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60)),
                        datetime.datetime(2019, 8, 7, 15, 37, 4, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=60))]


print('infer_dtype with offsets')
print('as list:',infer_dtype(data_with_offsets))
print('as Series:',infer_dtype(pd.Series(data_with_offsets)),'\n')

print('infer_dtype without offsets')
print('as list:',infer_dtype(data_without_offsets))
print('as Series:',infer_dtype(pd.Series(data_without_offsets)),'\n')

print('maybe_convert_objects (only works with numpy array) with offsets')
print('dtype='+str(maybe_convert_objects(np.array(data_with_offsets)).dtype),'\n')

print('maybe_convert_objects without offsets')
print('dtype='+str(maybe_convert_objects(np.array(data_without_offsets)).dtype),'\n')

print('is_datetime64tz_dtype (used in _parse_columns for read_sql) with offsets')
print('as list:',is_datetime64tz_dtype(data_with_offsets))
print('as Series:',is_datetime64tz_dtype(pd.Series(data_with_offsets)),'\n')

print('is_datetime64tz_dtype without offsets')
print('as list:',is_datetime64tz_dtype(data_without_offsets))
print('as Series:',is_datetime64tz_dtype(pd.Series(data_without_offsets)),'\n')

print('_is_datetime_series_with_different_offsets with offsets')
print(_is_datetime_series_with_different_offsets(pd.Series(data_with_offsets)),'\n')

print('_is_datetime_series_with_different_offsets without offsets')
print(_is_datetime_series_with_different_offsets(pd.Series(data_without_offsets)))
```
```
infer_dtype with offsets
as list: datetime
as Series: datetime 

infer_dtype without offsets
as list: datetime
as Series: datetime64 

maybe_convert_objects (only works with numpy array) with offsets
dtype=object 

maybe_convert_objects without offsets
dtype=object 

is_datetime64tz_dtype (used in _parse_columns for read_sql) with offsets
as list: False
as Series: False 

is_datetime64tz_dtype without offsets
as list: False
as Series: True 

_is_datetime_series_with_different_offsets with offsets
True 

_is_datetime_series_with_different_offsets without offsets
False
```

</details>

Reply:
pls don't re-invent the wheel, @mroeschke is correct here, if infer_dtype is not doing its job exactly right then then let's fix that, rather than creating an all new *mostly* untested code path.

Reply:
If we want to change infer_dtype we'd have to add an additional possible return value such as "datetimetz" because it currently does not differentiate between naive and timezone aware datetimes (see example below). And I don't think we want to convert postgres naive datetimes into UTC.

```python
from pandas.api.types import infer_dtype
import datetime

assert infer_dtype([datetime.datetime(2019,1,1,0,0,0, tzinfo = datetime.timezone.utc)]) == 'datetime'
assert infer_dtype([datetime.datetime(2019,1,1,0,0,0)]) == 'datetime'
```
Should we do that? Otherwise I see 2 other possibilities:

1. change is_datetime64tz_dtype so it returns True even with different offsets
2. simplify the logic in _is_datetime_series_with_different_offsets. This would be sufficient (if we do this we could get rid of the function and do it directly in _parse_date_columns):

```python
if infer_dtype(df_col) == 'datetime' and df_col.dtype == 'object':
    return True
```

Reply:
i would rather infer_dryoe return a new type (option 1)

we have extensive test coverage in this routine so adding this would be ok

Reply:
Yeah I think `infer_dtype` should be able to return `"datetimetz"`, somewhere around here.
https://github.com/pandas-dev/pandas/blob/7ab73a93c94e6557e5d77b2966549955e5c51cbc/pandas/_libs/lib.pyx#L1290

Reply:
I have made a new commit with an update on infer_dtype (it now separates between datetime and datetimetz). It seems like quite a big change (infer_dtype is used in lots of places).

I had to modify a method of SQLTable : _sqlalchemy_type to include this "datetimetz". Also in the same function the error message when using to_sql with different offsets/timezones changed so I had to kinda reeinvent the wheel there :crying_cat_face: